### DISCUSSION TITLE
LIS2DH12TR

### DISCUSSION THREAD

**Discussion by 0x0fe** [stm32duino/Arduino_Core_STM32#2273] [Libraries] [answered]
Is it possible to add LIS2DH12TR library for legacy support?

**Comment by fpistm** [comment]
Hi @0x0fe 
I let @cparata comment on this.

**Comment by 0x0fe** [comment]
ok thanks.

**Comment by 0x0fe** [comment]
also, side question (but fully related) does the SPI implementation uses DMA? It seems required to reach the max ODR of 5.3KHz on LIS2DH12, at least that is how i did it on SW4STM32 with the HAL at the time.

**Comment by cparata** [comment, marked as answer]
Hi @0x0fe ,
currently, the Arduino library to support the LIS2DH12 is not planned. Any contribution from the community will be appreciated. All the Arduino MEMS libraries are based on SPI calls in polling mode. This is because all the APIs are implemented in synchronous mode. The usage of DMA would require an abstraction layer for DMA stuff (to have the compatibility with all STM32 families) that at the moment, as I know, is not present in Arduino and also the adoption of APIs in asynchronous mode that is out of the scope of the Arduino MEMS libraries. For your use case, probably the best thing is to use the FIFO, that allow to store data without losses.
Best Regards,
Carlo

**Comment by 0x0fe** [comment]
@cparata yes i checked and notice none of the core drivers were implemented with DMA, anyway it turned out it was not required to read the datas at max ODR and so i used the standard polling SPI api.
Yes i wondered about using the FIFO to reduce the number of calls, but are samples guaranteed to be contiguous when using fifo? i dont see any double buffering mechanism. In my case it is mandatory to have contiguous samples because FFT and other processings are performed on the datas, in the previous implementation i polled sample by sample, checking the data  ready flag, so i did the same here and i can get all contiguous datas in time wiht no problem.  There is still something wrong because the FFT result differs from the previous implementation (with HAL), investigation is ongoing.

**Comment by cparata** [comment]
Hello @0x0fe ,
yes, the values of the FIFO are contiguous. Actually, the FIFO, when used in STREAM mode, will continue to buffer the sensor data for you. So, typically, you can set a threshold and start to empty the FIFO when you reach the wanted threshold. Then, you can use the read values to run your FFT algorithm; meantime the sensor will continue to fill the FIFO. If the process time of your algorithm is less than the time used by the sensor to fill the FIFO up to the threshold, you should not have issues. Obviously, you need to find a good compromise among output data rate of the sensor, FIFO threshold and execution time of your FFT algorithm in order to get everything working properly.
Best Regards,
Carlo

**Comment by 0x0fe** [comment]
@cparata thank you for confirming this, that will be convenient to perform acquisitions during the DSP time, and reduce the interrupts rate by a factor of 16.
